### ISSUE TITLE
Hows do you install use tfs-cli in a closed network environment?

### ISSUE THREAD

**Issue by jlongo62** [Microsoft/tfs-cli#94]
Our tfs build server does not have access to the internet. After getting node.js installed next I tried to install tfx-cli. using the command: npm install -g tfx-cli As far as I can tell, now it wants to download dependencies required. It looks like a whole web of dependencies - nuget style. Am I expected to set up an internal npm server with the dozens of dependencies required, becoming an expert on node.js and npm - just so I can add some custom commands to our tfs server ? Is there a version of tfs-cli I can get that includes all dependencies ?

This is a real problem - DoD, Government, Financial businesses, etc. do not want to expose development environments to the internet. A tool like tf.exe is a known quantity, installed from a setup utility. This npm/node.js - while limited to javascript vulnerabilities - requires internet access just to set up and is too hard to fly in controlled environments - and they are getting more and more strangling by the day.

**Comment by hounddog22030** [comment]
Joe?

**Comment by bryanmacfarlane** [comment]
I's actually not a problem.

npm install supports installing from a directory.  So, acquire from an internet connected machine and now you have a disconnected installer.  npm install ./<dir> -g

I can write more detailed docs later.  we could also "release" a zip here.

**Comment by jlongo62** [comment]
Who is this ?

**Comment by bryanmacfarlane** [comment]
The bottom of this page shows how to install from a folder (it installs from the _build/app folder)

https://github.com/Microsoft/tfs-cli/blob/master/docs/contributions.md

**Comment by jlongo62** [comment]
Your reply  is cryptic…

 

If you are trying to answer my question, on deploying tfx to a network isolated machine, please do. I neither am nor want to be an npm/node.js expert. I have already wasted a lot of time going down another open source rabbit hole – npm box; it does not work.

 

The problem is tfx has a whole tree of dependencies. They all need to packaged into the install. If you want to use npm to install, fine – but give us clear instructions. We are not experts at these scripting toys.

 

 

Regards,

Joseph Longo
.Net Healthcare Developer
703-819-8048
703-670-7026

**Comment by hounddog22030** [comment]
Classic Joe... Biting the hand that tries to feed you...

I suggest using Sonatype Nexus

**Comment by jlongo62** [comment]
Where do I know you from ?

 

Regards,

Joseph Longo
.Net Healthcare Developer
703-819-8048
703-670-7026

**Comment by hounddog22030** [comment]
Do you recognize the phrase "blocks of code decorated with the testmethod
attribute"?

**Comment by jlongo62** [comment]
no

 

Regards,

Joseph Longo
.Net Healthcare Developer
703-819-8048
703-670-7026

**Comment by jlongo62** [comment]
Your reply is cryptic…



If you are trying to answer my question, on deploying tfx to a network
isolated machine, please do.


The problem as I see it is tfx has a whole tree of dependencies. node.js and npm are designed by Open Source community that does not have to deal with closed environments – (those children  probably don’t even know what a closed environment is or why anyone would have one). The old tf.exe was a standalone, installed with a tfs client. Tfx requires node.js, then npm, then a slew of dependencies, and those dependencies begat other dependencies. If tfx is going to be the tool we need use going forward, it needs to be packaged into some kind of stand-alone kit. This package needs to include “clear” instructions for where to unzip it, what command to run, etc. A walk-though needs to be included that assumes the user does not have all the required pre-requisites.

 

Installing tfx was like peeling an onion. I have tfx running and now and I also know about npm and node, etc. – a huge waste of time. It took a long time to work out the details and was very frustrating. The “docs” assume the user already knows about nm and node.

 

I still don’t have tfx running on my build server – because I can’t download anything to it. Software has be transferred over from a separate machine. 

 

When working with Microsoft tools, I am used to not having  my time wasted by half-completed software. I hope Satya is not taking us into this world….

**Comment by kirkpabk** [comment]
Agree...   offline means offline...   it means that the consumer must be able to easily ascertain and gather the dependencies in a reasonable manner in order to support offline.  So far, no easy path has been defined.
